### PR TITLE
[fix](tvf) Partition columns in CTAS need to be compatible with the STRING type of external tables/TVF

### PR DESCRIPTION
**Pull request by BePPPower**
## Proposed changes

Issue Number: close #xxx

Partition columns in CTAS need to be compatible with the STRING type of external tables/TVF

1. For TVF queries with partition columns in the file path, map these partition columns to the `VARCHAR(65533)` type. (Previously, they were of `STRING` type.)

3. If you use the `STRING` type of an external table as a partition column, CTAS will automatically convert it to `VARCHAR(65535)` type."

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run p0 external

File: fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java [hunk 1]
```diff
@@ -1304,7 +1304,9 @@ public void createTableAsSelect(CreateTableAsSelectStmt stmt) throws DdlExceptio
                     }
                     if (resultExpr.getSrcSlotRef() != null
                             && resultExpr.getSrcSlotRef().getTable() != null
-                            && !resultExpr.getSrcSlotRef().getTable().isManagedTable()) {
+                            && !resultExpr.getSrcSlotRef().getTable().isManagedTable()
+                            && !createTableStmt.getPartitionDesc().getPartitionColNames().contains(
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java:1308]: getPartitionColNames() may return null

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run p0

**Review by morningman — APPROVED** [review]
LGTM

**Review by morningman — DISMISSED** [review]

**Comment by BePPPower** [comment]
run buildall

**Comment by BePPPower** [comment]
run feut

**Review by morningman — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/tablefunction/ExternalFileTableValuedFunction.java [hunk 2]
```diff
@@ -465,7 +465,7 @@ private void fillColumns(InternalService.PFetchTableSchemaResult result) {
         // HACK(tsy): path columns are all treated as STRING type now, after BE supports reading all columns
         //  types by all format readers from file meta, maybe reading path columns types from BE then.
         for (String colName : pathPartitionKeys) {
-            columns.add(new Column(colName, Type.STRING, false));
+            columns.add(new Column(colName, ScalarType.createVarcharType(65533), false));
```
Comment (wuwenchi) [inline fe/fe-core/src/main/java/org/apache/doris/tablefunction/ExternalFileTableValuedFunction.java:468]: replace `ScalarType.createVarcharType(65533)` with `VarcharType.MAX_VARCHAR_TYPE`

**Comment by BePPPower** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Review by wuwenchi — APPROVED** [review]

**Comment by BePPPower** [comment]
run p0

**Comment by BePPPower** [comment]
run p0

**Comment by BePPPower** [comment]
run cloud_p1

**Pull request merged**